### ISSUE TITLE
Potential boolean bug

### ISSUE THREAD

**Issue by pca006132** [elalish/manifold#970]
See https://github.com/elalish/manifold/pull/666#issuecomment-2356564020

Just to remind us not to forget about this after finishing v3.0 release.

**Comment by elalish** [comment]
Yeah, I feel like we could use an approximate `IsOverlapping` function  that we can run in debug mode after each Boolean so we can figure out where the problem first occurs, and dump out a test case of those meshes.

**Comment by pca006132** [comment]
For the record: We now check for self-intersection in https://github.com/elalish/manifold/blob/3a0c251b0e785022fcef36ac008143612e238680/src/properties.cpp#L213 and has a simple fuzzer in https://github.com/elalish/manifold/blob/master/test/manifold_fuzz.cpp that can fuzz this quickly.

**Comment by pca006132** [comment]
It seems that this may be related to the mesh simplification pass. For

https://github.com/elalish/manifold/blob/ac827679272a396972e92933c79602ba0aa496b9/test/boolean_complex_test.cpp#L1040-L1057

The result is not self-intersecting (doesn't even require epsilon-valid!) if we disable the mesh simplification pass, the part that flags redundant edges for collapsing. I think this suggests that the redundant edge removal may produce self-intersection.

However, disabling mesh simplification doesn't fix all the test cases. For example, the following case is still failing:

https://github.com/elalish/manifold/blob/ac827679272a396972e92933c79602ba0aa496b9/test/boolean_test.cpp#L425-L435

Probably two issues at play here.

---

Also, about the edge flagging code, why don't we do the following, which updates `ref1` to make sure `ref1` never points to the same triangle as `ref0`? The old code will not flag more edges, but it may flag less.

```c++
  bool operator()(int edge) const {
    if (halfedge[edge].pairedHalfedge < 0) return false;
    // Flag redundant edges - those where the startVert is surrounded by only
    // two original triangles.
    const TriRef ref0 = triRef[edge / 3];
    int current = NextHalfedge(halfedge[edge].pairedHalfedge);
    TriRef ref1 = triRef[current / 3];
    bool ref1Updated = !ref0.SameFace(ref1);
    while (current != edge) {
      current = NextHalfedge(halfedge[current].pairedHalfedge);
      int tri = current / 3;
      const TriRef ref = triRef[tri];
      if (!ref.SameFace(ref0) && !ref.SameFace(ref1)) {
        if (ref1Updated) {
          ref1 = ref;
          ref1Updated = true;
        } else {
          return false;
        }
      }
    }
    return true;
  }
```

**Comment by elalish** [comment]
Yeah, I've been poking around these functions, and they definitely need work. Care to put this in the form of a PR? I'll be able to give better feedback there.

**Comment by pca006132** [comment]
I think I found a case where it is giving invalid polygon to the triangulator:

```c++
  Manifold result =
      Manifold::Tetrahedron() +
      Manifold::Tetrahedron().Rotate(0.001, -0.08472872823860228, 0.055910459615905288) +
      Manifold::Tetrahedron();
```

which reports

```
Triangulation failed! Precision = 1.002453054734738789e-12
Error in file: /home/pca006132/code/manifold/src/polygon.cpp (124): 'std::all_of(triangles.begin(), triangles.end(), [&vertPos, epsilon](const ivec3 &tri) { return CCW(vertPos[tri[0]], vertPos[tri[1]], vertPos[tri[2]], epsilon) >= 0; })' is false: triangulation is not entirely CCW!
Polygon 0 1.002453054734738789e-12 2
14
1 -1
1 1
1 1
-0.2012216382419809435 -0.201221638241980999
0.9995243062182277427 -0.4973257940042222769
1 1
1 -1
1 1
1 -1
1 -1
1 -1
1 -1
0.9995243062182277427 -0.4973257940042222769
0.9985409151864541588 -1
4
-1 -1
-1 -1
-0.9975491063034749306 -0.9975491063034749306
-0.9975491063034749306 -0.9975491063034749306
# ... 
show(array([
  [1, -1],
  [1, 1],
  [1, 1],
  [-0.2012216382419809435, -0.201221638241980999],
  [0.9995243062182277427, -0.4973257940042222769],
  [1, 1],
  [1, -1],
  [1, 1],
  [1, -1],
  [1, -1],
  [1, -1],
  [1, -1],
  [0.9995243062182277427, -0.4973257940042222769],
  [0.9985409151864541588, -1],
]))
show(array([
  [-1, -1],
  [-1, -1],
  [-0.9975491063034749306, -0.9975491063034749306],
  [-0.9975491063034749306, -0.9975491063034749306],
]))
```

Visualized:

![Image](https://github.com/user-attachments/assets/6fb043b5-cc45-4c40-b2f2-25596a88ab18)

Bottom-right corner:

![Image](https://github.com/user-attachments/assets/7a5f86dd-b7a1-4ea4-a4b2-20d40d1cf460)

**Comment by starseeker** [comment]
@pca006132 Would the next step be to isolate where the invalid polygon is coming from?

**Comment by elalish** [comment]
So, I just drew this horrible polygon out on a piece of paper - believe it or not, it actually is epsilon-valid, even as epsilon approaches zero. The triangulator is choking on it though, so that's the bug in this case. Ideally the symbolic perturbation would work better to keep so much degeneracy from being formed, but I think it's pretty inevitable, especially with sharp corners like the tetrahedron, where the normal won't necessarily move every face in the right direction.

**Comment by pca006132** [comment]
Yeah, the behavior is changed after making things deterministic. In fact, I can only produce the problematic polygon with 1 in 10 runs, roughly.

**Comment by elalish** [comment]
Hmm, I guess that means it's not quite deterministic yet? Any idea what's left?

**Comment by pca006132** [comment]
Sorry I was not being accurate here. I can only reproduce it like that before making things deterministic. Currently it is in a code path that never triggers the behavior.
